lib/src/wrappers.dart

```diff
@@ -469,13 +469,33 @@ class DelegatingMap<K, V> implements Map<K, V> {
 /// Note that [lookup] is not supported for this set.
 class MapKeySet<E> extends _DelegatingIterableBase<E>
     with UnmodifiableSetMixin<E> {
+  /// Adapts [source] to be a `MapKeySet<T>`.
+  ///
+  /// Any time the set would produce an element that is not a [T], the element
+  /// access will throw.
+  ///
+  /// Any time a [T] value is attempted stored into the adapted set, the store
+  /// will throw unless the value is also an instance of [S].
+  ///
+  /// If all accessed elements of [source] are actually instances of [T] and if
+  /// all elements stored in the returned  are actually instance of [S],
+  /// then the returned set can be used as a `MapKeySet<T>`.
+  static MapKeySet<T> castFrom<S, T>(MapKeySet<S> source) {
+    return new MapKeySet(source._baseMap.cast<T, dynamic>());
+  }
+
   final Map<E, dynamic> _baseMap;
 
   MapKeySet(Map<E, dynamic> base) : _baseMap = base;
 
   Iterable<E> get _base => _baseMap.keys;
 
-  Set<T> cast<T>() => _baseMap.keys.toSet().cast<T>();
+  MapKeySet<T> cast<T>() {
```

Comment:
Just return `Set<T>` and use `Set.castFrom`.

Reply:
Done.